File: python/백준/자료구조/스택, 집합/1918 후위 표기식.py
string = input()

result = []
stack = []


for c in string:
    if c in ('+', '-', '*', '/'):
        if c in ('+', '-'):
            while stack and stack[-1] != '(' and stack[-1] in ('+','-','*','/'):
                result.append(stack.pop())
            stack.append(c)
        else:  # c in ('*','/')
            while stack and stack[-1] != '(' and stack[-1] in ('*','/'):
                result.append(stack.pop())
            stack.append(c)
    
    elif c == '(':
        stack.append('(')
        
    elif c == ')':
        while stack and stack[-1] != '(':
            result.append(stack.pop())
        stack.pop()
        
    else: 
        result.append(c)
    
while stack:
    result.append(stack.pop())
        
print("".join(result))

File: cpp/백준/자료구조/스택, 집합/1918 후위 표기식.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    string s;
    cin >> s;

    string result;
    stack<char> st;

    for (char c : s) {
        if (c == '+' || c == '-') {
            while (!st.empty() && st.top() != '(' &&
                   (st.top() == '+' || st.top() == '-' || st.top() == '*' || st.top() == '/')) {
                result += st.top();
                st.pop();
            }
            st.push(c);
        } else if (c == '*' || c == '/') {
            while (!st.empty() && st.top() != '(' &&
                   (st.top() == '*' || st.top() == '/')) {
                result += st.top();
                st.pop();
            }
            st.push(c);
        } else if (c == '(') {
            st.push('(');
        } else if (c == ')') {
            while (!st.empty() && st.top() != '(') {
                result += st.top();
                st.pop();
            }
            st.pop();
        } else {
            result += c;
        }
    }

    while (!st.empty()) {
        result += st.top();
        st.pop();
    }

    cout << result << "\n";
    return 0;
}
